src/Device/Driver/AR62xx.cpp

```diff
@@ -0,0 +1,443 @@
+    active_frequency_idx = FrequencyToAR62xxId(frequency);
+  }
+
+  /* set passive station */
+  else
+  {
+    passive_frequency = frequency;
+    passive_frequency_idx = FrequencyToAR62xxId(frequency);
+  }
+
+  /* setting frequencies */
+  command[command_length++] = 22;
+
+  //add active station
+  uint8_t tmp[2] = {};
+  WriteUnalignedLE16((uint16_t *)(void *)tmp, active_frequency_idx);
+  command[command_length++] = tmp[1];
+  command[command_length++] = tmp[0];
+
+  //add passive station
+  WriteUnalignedLE16((uint16_t *)(void *)tmp, passive_frequency_idx);
+  command[command_length++] = tmp[1];
+  command[command_length++] = tmp[0];
+
+  //add crc
+  WriteUnalignedLE16((uint16_t *)(void *)tmp, CRCBitwise(command, command_length));
+  command[command_length++] = tmp[1];
+  command[command_length++] = tmp[0];
+
+  //return full command
+  return command_length;
+}
+
+bool AR62xxDevice::ParseReceivedData(const uint8_t *data, size_t len)
+{
+  size_t idx = 0;
+  size_t rec_buf_len = 0;
+
+  /* we only need to read the message until record-size = 10 because then it already contains
+   * all data we want to read (active and passive frequency) */
+  size_t rec_bufsize = 10;
+  uint8_t command[rec_bufsize];
+
+  //flag if frequency has been saved
+  bool frequencySaved = false;
+
+  /* loop as long we have data and stop if frequency has been set */
+  while (!frequencySaved && rec_buf_len < rec_bufsize && idx < len)
+  {
+    /* check if we received a header which tells us now data starts */
+    if (data[idx] == 0xA5)
+    {
+      /* reset buffer length */
+      rec_buf_len = 0;
+    }
+
+    /* write command in buffer */
+    command[rec_buf_len++] = (char)data[idx++];
+
+    /* check if record-length == 10 then we can read the frequencies (active and passive frequency of ar62xx)
+     */
+    if (rec_buf_len == rec_bufsize)
+    {
+      size_t command_length = command[2];
+
+      /* generate uint16_t out of commands */
+      u_int8_t tmp[] = {command[command_length + 4], command[command_length + 3]};
+      uint16_t crc_value = ReadUnalignedLE16((const uint16_t *)(const void *)tmp);
+
+      /* generate crc */
+      uint16_t crc = CRCBitwise(command, command_length + 3);
+
+      /* check crc */
+      if (crc == crc_value || crc_value == 0)
+      {
+        /* extract and save active and passive frequency */
+        SaveAR62xxFrequencies(command, command_length + 5, crc);
+        frequencySaved = true;
+      }
+    }
+  }
+  return true;
+}
+
+void AR62xxDevice::SaveAR62xxFrequencies(uint8_t *command, size_t len, uint16_t crc)
+{
+  /* command 22 = frequency-settings
+   *  Frequency settings, always for both frequencies (active and passive) */
+  if ((unsigned char)(command[3] & 0x7F) == 22)
+  {
+    u_int8_t tmp_active[] = {command[5], command[4]};
+    active_frequency = AR62xxIdToFrequency(ReadUnalignedLE16((const uint16_t *)(const void *)tmp_active));
+
+    u_int8_t tmp_passive[] = {command[7], command[6]};
```

Comment:
Yet another useless stack copy

Reply:
solved... please have a look at the new implementation which is completely different from this and fully documented :-)